Add alert actions api

moolitayer opened the pull request:
Adds retrieval and creation of alert_actions as a sub resource of alerts.

Route | Body
--- | ---
GET /api/alerts/:id/alert_actions/ | [1]
GET /api/alerts/:id/alert_actions/:id | [2]
POST /api/alerts/:id/alert_actions/ | [3]

[1] :
Response:
```
{
    "name": "alert_actions",
    "count": 3,
    "subcount": 3,
    "resources": [
        {
            "href": "http://localhost:3000/api/alerts/6/alert_actions/8"
        },
        {
            "href": "http://localhost:3000/api/alerts/6/alert_actions/9"
        },
        {
            "href": "http://localhost:3000/api/alerts/6/alert_actions/10"
        }
    ],
    "actions": [
        {
            "name": "create",
            "method": "post",
            "href": "http://localhost:3000/api/alerts/6/alert_actions"
        }
    ]
}
```

[2] :
Response:
```
{
    "href": "http://localhost:3000/api/alerts/6/alert_actions/9",
    "id": 9,
    "action_type": "assign",
    "user_id": 1,
    "assignee_id": 1,
    "miq_alert_status_id": 6,
    "created_at": "2017-01-22T12:29:29Z",
    "updated_at": "2017-01-22T12:29:29Z"
}
```

[3] :
Request:
```
{
  "action_type": "comment",
  "user_id": 1,
  "comment": "comment text"
}
```
Response:
```
{
    "results": [
        {
            "id": 11,
            "action_type": "comment",
            "user_id": 1,
            "comment": "comment text",
            "miq_alert_status_id": 6,
            "created_at": "2017-01-22T12:36:54Z",
            "updated_at": "2017-01-22T12:36:54Z"
        }
    ]
}

moolitayer commented:
@miq-bot add_label wip, enhancement, api

moolitayer commented:
@miq-bot remove_label wip

moolitayer commented:
@abellotti can you start reviewing 530fdbb9e46ccc7734ded66aa05e8b5afb93c8c7 even though the dependent patch hasn't been merged yet? 

It's critical to have all of those patches merged by the end of the week to have the API in place for the next week for other people's work

moolitayer commented:
@abellotti PTAL

abellotti commented:
Could you add a description entry showing the expected usage, i.e.

```
GET /api/alerts
GET /api/alerts/:id/alert_actions
POST /api/alerts/:id/alert_actions  with action names ...
...
```

It would help us to better review.

abellotti commented on app/controllers/api/subcollections/alerts.rb:
```diff
@@ -0,0 +1,9 @@
+  module Subcollections
+    module Alerts
+      def alerts_query_resource(object)
+        object.try(:miq_alert_statuses).to_a
```
could just be:

```
object.miq_alert_statuses || []
```

abellotti commented on config/api.yml:
```diff
@@ -647,6 +647,46 @@
+       - :name: read
+         :identifier: alert_status_action_show_list
+       :post:
+       - :name: create
```
does it make sense to allow creates as both /api/alert_actions as well as /api/alerts/:id/alert_actions ?  Doesn't the former need a relationship to the /api/alerts resource ?

moolitayer replied:
Right, should be supported only for /api/alerts/:id/alert_actions now

abellotti commented on spec/requests/api/alert_actions_spec.rb:
```diff
@@ -0,0 +1,36 @@
+    expect(response.parsed_body["results"].count).to eq(1)
+    expect(response.parsed_body["results"][0]).to include(specification)
+  end
+  # TODO: add tests for sub collection (only related for alert returned, id is ignored and taken from parent etc)
```
we'd need this test added too.

abellotti commented on spec/requests/api/alert_actions_spec.rb:
```diff
@@ -0,0 +1,36 @@
+    run_post(alert_actions_url, specification)
+    expect(response).to have_http_status(:ok)
+    expect(response.parsed_body["results"].count).to eq(1)
+    expect(response.parsed_body["results"][0]).to include(specification)
```
32/33 could be a single expect.

abellotti commented on db/fixtures/miq_product_features.yml:
```diff
@@ -1932,7 +1932,41 @@
+  - :name: Add
+    :description: Display Individual Alert Status
+    :feature_type: control
+    :identifier: alert_status_action_new
```
/cc @h-kataria need your 👀 on above miq_product_features.yml changes.  Thanks!!

moolitayer commented:
> /cc @h-kataria need your  on above miq_product_features.yml changes. Thanks!!

@h-kataria could you please take a look at the product features for added here?
Note this is part of the API needed for "Monitor" added [here](https://github.com/ManageIQ/manageiq/pull/13357/files) does it make sense to have separate features?

@imtayadeway @abellotti addressed comments, and made some changes based on the review of #13025 can you take a look? Thanks

moolitayer commented:
@imtayadeway @abellotti Please review, Feedback appriciated

imtayadeway commented on app/controllers/api/subcollections/alert_actions.rb:
```diff
@@ -0,0 +1,19 @@
+  module Subcollections
+    module AlertActions
+      def alert_actions_query_resource(object)
+        MiqAlertStatusAction.where(:miq_alert_status_id => object.id)
```
Do alert statuses not have alert status actions? i.e. something like `object.miq_alert_status_actions`?

moolitayer replied:
Sure, will change

imtayadeway commented on app/controllers/api/subcollections/alert_actions.rb:
```diff
@@ -0,0 +1,19 @@
+      end
+
+      def alert_actions_create_resource(object, type, _id, data)
+        data['miq_alert_status_id'] = object.id
```
This is changing the data that's passed into this method - perhaps you could `dup` the data before you do this?

moolitayer replied:
Maybe using merge would be cleaner?

imtayadeway commented on spec/requests/api/alerts_spec.rb:
```diff
@@ -44,4 +44,97 @@
+      api_basic_authorize subcollection_action_identifier(:alerts, :alert_actions, :create, :post)
+      run_post(actions_subcollection_url, attributes)
+      expect(response).to have_http_status(:ok)
+      alert_action = MiqAlertStatusAction.find(response.parsed_body["results"].first["id"])
```
Are the attributes not sent back in the response? Can you not put the expectation on those?

moolitayer commented:
@imtayadeway addressed comments

imtayadeway commented on spec/requests/api/alerts_spec.rb:
```diff
@@ -44,4 +44,96 @@
+      api_basic_authorize subcollection_action_identifier(:alerts, :alert_actions, :create, :post)
+      run_post(actions_subcollection_url, attributes)
+      expect(response).to have_http_status(:ok)
+      expect(response.parsed_body["results"].first).to have_attributes(attributes)
```
This is fine, but in the case that it breaks you may get a NoMethodError on `nil`, instead of a diff. So it may be better to write these and others like it something like:


expected = {
  "results" => [
    a_hash_including(attributes)
  ]
}
expected(response.parsed_body). to include(expected)
```

imtayadeway commented on spec/requests/api/alerts_spec.rb:
```diff
@@ -44,4 +44,96 @@
+        "comment"     => "comment text",
+      )
+      expect(response).to have_http_status(:bad_request)
+      expect(response.parsed_body["error"]["message"]).to eq(
```
There's a helper that you can use here that will better handle the case outlined above:

```ruby
expect(response).to have_error_with_message("Failed etc...")
```

imtayadeway commented:
@moolitayer this looks good, just some minor nits about tests. @abellotti any thoughts?

moolitayer commented:
Thanks @imtayadeway!
Addressed comments, @abellotti please review

Good night :moon: :zzz:

abellotti commented on app/controllers/api/subcollections/alert_actions.rb:
```diff
@@ -0,0 +1,20 @@
+        )
+        if alert_action.invalid?
+          raise BadRequestError, "Failed to add a new alert action resource" \
+                                 " - #{alert_action.errors.full_messages.join(', ')}"
```
would prefer:

```
          raise BadRequestError,
                "Failed to add a new alert action resource - #{alert_action.errors.full_messages.join(', ')}"
 ```

should still be good rubucop line length limit-wise.

abellotti commented on spec/requests/api/alerts_spec.rb:
```diff
@@ -44,4 +44,101 @@
+      expect(response).to have_http_status(:forbidden)
+    end
+
+    it "creats an alert action under an alert" do
```
creats -> creates

abellotti commented on spec/requests/api/alerts_spec.rb:
```diff
@@ -44,4 +44,101 @@
+      user = FactoryGirl.create(:user)
+      attributes = {
+        "action_type" => "comment",
+        "user_id"     => user.id,
```
should support user reference signatures in addition to user_id,

i.e. support:

```
   "user" : { "id" : ### }
```

as well as:

```
  "user" : { "href" : "http://localhost:3000/api/users/###" }
```

abellotti replied:
and the related tests for both additional signatures.

moolitayer replied:
OK I Hope I understood. I should not accept a parameter named user_id anymore?

abellotti replied:
keep support for user_id as you do today (that's what is sent to the model), but the API should support the user reference and generate the user_id accordingly, see what was done for arbitration_rules creation for specifying arbitration_profile  (we define the arbitration_profile_id if the arbitration_profile reference is specified, using parse_id) https://github.com/ManageIQ/manageiq/pull/13525/files

moolitayer replied:
Thanks for the clarification.

Now that I think about it
- the user should always be the current user
- your comment should be applied on assignee_id

user_id === who made this action
assignee_id === assignment target, if this is an assignment 

I'm working on it, let me know if that does not make sense

abellotti replied:
if user_id is always who created this action, why is it passed in below as part of attributes in the run_post (line 102).  Maybe it should not and the API forces it to be User.current_user.id before sending to model ?

moolitayer replied:
That is should be the situation now. I wrote a test to show a user_id is always ignored in favor of the current user ("creates an alert action on the current user")

abellotti commented on spec/requests/api/alerts_spec.rb:
```diff
@@ -44,4 +44,101 @@
+        "subcount"  => 1,
+        "resources" => [
+          {
+            "href" => "http://www.example.com/api/alerts/#{alert.id}/alert_actions/#{alert_action.id}"
```
would prefer:

```
  "resources" = [
     {
         "href" => a_string_matching("#{alerts_url(alert.id)}/alert_actions/#{alert_action.id}")
     }
```

including the ones in the alerts_spec.rb, but those can be a different PR.

abellotti commented on spec/requests/api/alerts_spec.rb:
```diff
@@ -44,4 +44,101 @@
+      run_get("#{actions_subcollection_url}/#{alert_action.id}")
+      expect(response).to have_http_status(:ok)
+      expect(response.parsed_body).to have_attributes(
+        "href"        => "http://www.example.com/api/alerts/#{alert.id}/alert_actions/#{alert_action.id}",
```
same as above, would prefer the a_string_matching, leveraging alerts_url.

moolitayer commented:
@abellotti addressed comments, added tests for the new functionalities
(assigne can be refrenced by either id or href, user value should be ignored and always set to the current user)
 please take a look

abellotti commented:
LGTM!! Thanks @moolitayer for the API Enhancement.

moolitayer commented:
Awesome @abellotti @imtayadeway Thanks for reviewing & improving my work
